kubernetes/skyring/templates/statefulset.yaml

```diff
@@ -0,0 +1,170 @@
+            - name: seeds
+              value: '{{ .Values.nameOverride }}-seed:3455'
+            {{- if not .Values.persistent.enabled }}
+            - name: storage__backend
+              value: 'leveldown'
+            - name: storage__path
+              value: /var/data/skyring
+            {{- end }}
+            {{- if .Values.persistent.enabled }}
+            - name: storage__backend
+              value: 'mongodown'
+            - name: storage__path
+              value: 'mongodb://{{ .Release.Name }}-mongodb:27017/skyring'
+            - name: storage__collection
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+            {{ end }}
+          resources:
+{{ toYaml .Values.resources | indent 12 }}
+      {{- if .Values.nodeSelector }}
+      nodeSelector:
+{{ toYaml .Values.nodeSelector | indent 8 }}
+    {{- end }}
+    {{- if .Values.tolerations }}
+      tolerations:
+{{ toYaml .Values.tolerations | indent 8 }}
+    {{- end }}
+---
+apiVersion: apps/v1beta2
+kind: StatefulSet
+metadata:
+  name: {{ .Values.nameOverride }}
+  labels:
+    app: {{ .Values.nameOverride }}
+    chart: {{ template "skyring.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  replicas: {{ .Values.WorkerReplicaCount }}
+  selector:
+    matchLabels:
+      app: {{ .Values.nameOverride }}
+      release: {{ .Release.Name }}
+  template:
+    metadata:
+      labels:
+        app: {{ .Values.nameOverride }}
+        release: {{ .Release.Name }}
+    spec:
+      initContainers:
+      - name: nats-pause
+        image: yauritux/busybox-curl
+        command:
+        - curl
+        - "http://{{ .Release.Name }}-nats-client:4222/monitoring"
+      {{- if .Values.persistent.enabled }}
+      - name: mongo-pause
+        image: yauritux/busybox-curl
+        command:
+        - curl
+        - "http://{{ .Release.Name }}-mongodb-headless:9216/metrics"
+      {{- end }}
+      containers:
+        - name: {{ .Values.nameOverride }}
+          image: {{ .Values.image.repository }}:{{ if not .Values.persistent.enabled }}{{ .Values.image.tag }}{{ end }}{{ if .Values.persistent.enabled }}{{ .Values.image.persistent_tag }}{{ end}}
+          imagePullPolicy: Always
+          ports:
+            - name: port-one
+              containerPort: 3455
+            - name: port-two
+              containerPort: 3456
+            - name: port-three
+              containerPort: 3000
+          env:
+            {{- if eq .Values.debug true }}
+            - name: DEBUG
+              value: "*"
+            {{- end}}
+            - name: channel__host
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+            - name: nats__hosts
+              value: '{{ .Release.Name }}-nats-client:4222'
+            - name: seeds
+              value: '{{ .Values.nameOverride}}-seed:3455'
```

Comment:
@esatterwhite Yes absolutely makes sense. This was just a quick mock docker compose I made the error was in contactPoints. I need a quay image that has the mongo dependency and a quay image that has the scylladb dependency then I can make a pr that has leveldown w/ persistent storage, mongo, and scylladb support. 

I tested all three of them fairly heavily using my own images and they are really resilient to failures. 

There are a few bugs however but these are fine:

1. When using mongo for persistency if the Mongo Primary dies it can not recover as fast as needed - if timers try to clear or create while Mongo primary is down the node will crash

2. If a node dies while handling a large amount of timers a race condition will occur between timers firing and timers being cleared out as it passes off the timers to other nodes during shutdown. Since Kubernetes clusters can be run across multiple datacenters and nodes I propose the option to not pass off the timers at all during shutdown, I made these changes on a fork and tested it and this fits the Kubernetes structure much better. If a node dies that same pod can appear on another instance with the same Persistent Volume and recovery will bring back all the timers quickly and accurately. Furthermore, if a node crashes in the first place it often was unable to pass off timers anyways, so I felt that just standardizing on crash-like scenarios for Kubernetes always preserved the timers and their accuracy as pods can spin up in under 1-2 seconds after a crash for little to no downtime. 

3. Very rarely when a node dies - crash or regular shutdown, timers will RARELY be repeated. In a scenario of creating 1,000 timers a minute for 10 seconds every second and killing a node in the middle of this roughly 1 timer would fire twice every second trial. This is due to a race condition in the timer firing and clearing out persistent storage, I can create a pr for this as I identified where exactly this is.

So to summarize: When nodes don't pass off timers and just use the leveldown recovery the timers are incredibly resilient and due to Kubernetes speed of being able to spin up the same pods after a crash in a different datacenter these timers will always fire at accurate times.

Reply:
thanks for the detailed report. 
It sounds mostly like some work needs to go into the rebalance operations to make sure timers aren't executed when they are being moved. It sounds like the meat of the problem.

And you're right. mongo probably isn't the best choice. I originally set up mongo as an example - that remote storage is possible ( and people love mongo ).

a local leveldb is always going to be the best option.
scylla is going to be a pretty close 2nd if you really want to off load storage.

I'll be working on #27 for a while but the rebalance stuff is next on the list.

Reply:
you could add the two additional package to the docker image

https://github.com/esatterwhite/skyring/blob/master/Dockerfile#L18

```
npm install && npm install mongodown @skyring/scylladown --no-save ...
```